docs/proposals/policy.md

```diff
@@ -0,0 +1,146 @@
+// are paired with Authorizers during registration.
+type AuthorizationAttributeBuilder interface{
+	GetAttributes(request http.Request) (AuthorizationAttributes, error)
+}
+
+type AuthorizationAttributes interface{
+	GetUserInfo() UserInfo  // includes GetName() string GetGroups() []string
+	GetVerb() string
+	GetNamespace() string
+	// GetRequestAttributes is of type interface{} because different verbs and different 
+	// Authorizer/AuthorizationAttributeBuilder pairs may have different contract requirements
+	GetRequestAttributes() interface{}
+}
+```
+A single Authorizer/AuthorizationAttributeBuilder pair may be registered with the apiserver.
+
+
+### Verb Set
+Currently, we only have "readonly" and everything else.  This does not give sufficient control.  We need to support all of the standard HTTP verbs, plus different concepts such as "watch" that don't have equivalents.  Our authorizer will start with the following verbs, but other implementations could choose different ones:
+  1.  exec
+  2.  watch
+  3.  list
+  4.  get
+  5.  create
+  6.  update
+  7.  delete
+  8.  proxy
+
+
+### Verb Specific Attributes
+Different verbs will require different attributes.  For instance, the exec verb will have information about the command being run, while a get will have the resource kind being requested, and an update may have a list of the fields being modified.  The exact contract rules will be chosen by the Authorizer/AuthorizationAttributeBuilder pair, but basic resource access for openshift should include  kind.
+
+
+### Configtime Authorization Types
+```
+// PolicyRule holds information that describes a policy rule, but does not contain information 
+// about who the rule applies to or which namespace the rule applies to.  This makes PolicyRules portable units.
+type PolicyRule struct {
+	// Deny is true if any request matching this rule should be denied.  If false, any request matching this rule is allowed.
+	Deny bool
+	// Verbs is a list of Verbs that apply to ALL the ResourceKinds and AttributeRestrictions contained in this rule
+	Verbs []Verb
+	// AttributeRestrictions will vary depending on what the Authorizer/AuthorizationAttributeBuilder pair supports.  
+	// Making it an interface{} allows a type switch for identification followed by properly typed usage thereafter.
+	AttributeRestrictions interface{}
+	// ResourceKinds is a list of kinds this rule applies to
+	ResourceKinds []string
+}
+
+// PolicyRuleBinding builds on top of a PolicyRule.  It adds who information (usernames, teamnames, accountnames) 
+// and namespace information (ObjectMeta).  PolicyRuleBindings only apply to the namespace they are contained in.  
+// That means the full set of PolicyRuleBindings for a namespaced request are contained inside of the requested 
+// namespace or in the global namespace
+type PolicyRuleBinding struct {
+	kapi.TypeMeta
+	kapi.ObjectMeta
+
+	UserNames    []string
+	TeamNames    []string
+	AccountNames []string
```

Comment:
I was saying account names is redundant with user names - teams and groups and roles are functionally identical, so need to be thought of as distinct but similar.  Teams or something like them stays.

The goal of roles is to apply a consistent set of policy rules similar to how team abstracts a group of people.  You have the pieces of that, but I'm not sure that the distinct binding resources are necessary.

Use case: I want to apply the same set of policy and roles to large numbers of project. When a new project is created it automatically gets a policy and roles. As an admin, I can globally change the policy of a single role and have it applied to all projects using that policy.

Use case: I want to customize the policy of a project, and have that policy inherit the bulk of its logic from the original policy.  Is it acceptable for the policy to be copied (so now there are two policies)?  If an admin customizes that policy, it's likely they want to reuse that customized policy.

Open questions:

1. How do admins ensure the right policy is assigned on project creation?  Does a deployer of OpenShift have to wrap project creation?

2. When an admin customizes a policy to set roles, what are the odds they want to allow project admins to modify the policy?



> On Jan 2, 2015, at 2:14 PM, David Eads <notifications@github.com> wrote:
> 
> In docs/proposals/policy.md:
> 
> > +	AttributeRestrictions interface{}
> > +	// ResourceKinds is a list of kinds this rule applies to
> > +	ResourceKinds []string
> > +}
> > +
> > +// PolicyRuleBinding builds on top of a PolicyRule.  It adds who information (usernames, teamnames, accountnames) 
> > +// and namespace information (ObjectMeta).  PolicyRuleBindings only apply to the namespace they are contained in.  
> > +// That means the full set of PolicyRuleBindings for a namespaced request are contained inside of the requested 
> > +// namespace or in the global namespace
> > +type PolicyRuleBinding struct {
> > +	kapi.TypeMeta
> > +	kapi.ObjectMeta
> > +
> > +	UserNames    []string
> > +	TeamNames    []string
> > +	AccountNames []string
> user names should be the only option.
> 
> Account names I wasn't sure of, but teams seem like a good idea. It seems reasonable that someone would want to be able to easily describe a rule that says, "George can modify team memberships, but cannot modify the rules that govern what members of that team can do". It also seems reasonable that you'd want to be able to describe a re-useable group of people and grant permissions to that group of people without enumerating them every time.
> 
> —
> Reply to this email directly or view it on GitHub.
>